### PR TITLE
fix issue #48 - Redshift does not support sequences

### PR DESCRIPTION
**Pull request by mtambos**
## Todos
- [ ] MIT compatible
- [ ] Tests
- [ ] Documentation
- [ ] Updated CHANGES.rst

### CODE REVIEW

**Comment by jklukas** [comment]
Is there a test that could illustrate the problem this fixes? Is there any possibility this could break existing workflows?

**Comment by mtambos** [comment]
> Is there a test that could illustrate the problem this fixes?

The following script:

```
from sqlalchemy import Column, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy_redshift.dialect import RedshiftDialect

# These are the current (in sqlalchemy-redshift:master) settings
RedshiftDialect.supports_sequences = True
RedshiftDialect.preexecute_autoincrement_sequences = True
RedshiftDialect.postfetch_lastrowid = False


Base = declarative_base()


class Foo(Base):
    __tablename__ = 'foo'

    id = Column(Integer, primary_key=True)
    some_int = Column(Integer)

conn_str = 'redshift+psycopg2://{user}:{password}@{host}:{port}/{database}'
conn_str = conn_str.format(**{connection parameters})

engine = create_engine(conn_str)
with engine.begin():
    Base.metadata.create_all(engine)

    Session = sessionmaker(bind=engine)
    session = Session()

    try:
        foo = Foo(some_int=9)
        session.add(foo)
        print session.query(Foo).filter(Foo.some_int == 9).first()
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()
```

produces the following error (excerpt):

```
  File "[...]/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.StatementError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (sqlalchemy.exc.ProgrammingError) (psycopg2.ProgrammingError) relation "foo_id_seq" does not exist
 [SQL: 'select nextval(\'"foo_id_seq"\')'] [SQL: u'INSERT INTO foo (id, some_int) VALUES (%(id)s, %(some_int)s)'] [parameters: [{'some_int': 9}]]
```

> Is there any possibility this could break existing workflows?

Not that I'm aware, since the [documentation](http://docs.aws.amazon.com/redshift/latest/dg/c_unsupported-postgresql-features.html) explicitly states that Redshift does not support sequences:

**Comment by graingert** [comment]
Does postfetch_lastrowid = True work? psycopg doesn't support cursor.lastrowid() and the psycopg dialect does not override get_lastrowid().

**Comment by mtambos** [comment]
> Does postfetch_lastrowid = True work?
It seems to work.
If you run this with `RedshiftDialect.postfetch_lastrowid = True`:

```
    session = Session()
    try:
        for i in range(50, 5):
            session.add(Foo(some_int=i))
        foo2 = session.query(Foo).filter(Foo.some_int == 54).first()
        print foo2.id
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()
```
You get the row's id. However, if you run the same with `RedshiftDialect.postfetch_lastrowid = False`, this happens:
```
Traceback (most recent call last):
  File "test.py", line 44, in <module>
    print foo2.id
AttributeError: 'NoneType' object has no attribute 'id'

```

**Comment by graingert** [comment]
For my own personal interest because I thought this was impossible, can you
log the SQL this runs?

Also can you convert this into a test in this repo.
On 6 May 2016 15:02, "mtambos" <notifications@github.com> wrote:

Does postfetch_lastrowid = True work?
It seems to work.
If you run this with RedshiftDialect.postfetch_lastrowid = True:

    session = Session()
    try:
        for i in range(50, 5):
            session.add(Foo(some_int=i))
        foo2 = session.query(Foo).filter(Foo.some_int == 54).first()
        print foo2.id

        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()

You get the row's id. However, if you run the same with
RedshiftDialect.postfetch_lastrowid
= False, this happens:

Traceback (most recent call last):
  File "test.py", line 44, in <module>
    print foo2.id
AttributeError: 'NoneType' object has no attribute 'id'


—
You are receiving this because you commented.

Reply to this email directly or view it on GitHub
<https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/85#issuecomment-217448664>

**Comment by mtambos** [comment]
This is the output generated with `echo=True`:
```
2016-05-08 12:59:24,982 INFO sqlalchemy.engine.base.Engine select version()
2016-05-08 12:59:24,982 INFO sqlalchemy.engine.base.Engine {}
2016-05-08 12:59:25,327 INFO sqlalchemy.engine.base.Engine select current_schema()
2016-05-08 12:59:25,327 INFO sqlalchemy.engine.base.Engine {}
2016-05-08 12:59:25,670 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2016-05-08 12:59:25,670 INFO sqlalchemy.engine.base.Engine {}
2016-05-08 12:59:25,839 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2016-05-08 12:59:25,839 INFO sqlalchemy.engine.base.Engine {}
2016-05-08 12:59:26,353 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2016-05-08 12:59:28,102 INFO sqlalchemy.engine.base.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
2016-05-08 12:59:28,102 INFO sqlalchemy.engine.base.Engine {'name': u'foo'}
2016-05-08 12:59:28,457 INFO sqlalchemy.engine.base.Engine 
CREATE TABLE foo (
	id INTEGER IDENTITY(0,1) NOT NULL, 
	some_int INTEGER, 
	PRIMARY KEY (id)
)


2016-05-08 12:59:28,457 INFO sqlalchemy.engine.base.Engine {}
2016-05-08 12:59:28,692 INFO sqlalchemy.engine.base.Engine COMMIT
2016-05-08 12:59:28,969 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2016-05-08 12:59:28,969 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (some_int) VALUES (%(some_int)s)
2016-05-08 12:59:28,969 INFO sqlalchemy.engine.base.Engine {'some_int': 0}
2016-05-08 12:59:29,326 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (some_int) VALUES (%(some_int)s)
2016-05-08 12:59:29,326 INFO sqlalchemy.engine.base.Engine {'some_int': 1}
2016-05-08 12:59:29,508 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (some_int) VALUES (%(some_int)s)
2016-05-08 12:59:29,508 INFO sqlalchemy.engine.base.Engine {'some_int': 2}
2016-05-08 12:59:29,697 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (some_int) VALUES (%(some_int)s)
2016-05-08 12:59:29,697 INFO sqlalchemy.engine.base.Engine {'some_int': 3}
2016-05-08 12:59:29,882 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (some_int) VALUES (%(some_int)s)
2016-05-08 12:59:29,883 INFO sqlalchemy.engine.base.Engine {'some_int': 4}
2016-05-08 12:59:30,068 INFO sqlalchemy.engine.base.Engine SELECT foo.id AS foo_id, foo.some_int AS foo_some_int 
FROM foo 
WHERE foo.some_int = %(some_int_1)s 
 LIMIT %(param_1)s
2016-05-08 12:59:30,068 INFO sqlalchemy.engine.base.Engine {'some_int_1': 4, 'param_1': 1}
8

2016-05-08 12:59:30,249 INFO sqlalchemy.engine.base.Engine COMMIT
2016-05-08 12:59:30,537 INFO sqlalchemy.engine.base.Engine COMMIT
```

BTW, I ended up configuring the `id` column as:

```
id = Column(Integer, primary_key=True, info={'identity': (0, 1)})
```

Regarding the tests, I've pushed them to my fork, but after a couple hours I haven't been able to get them to work.
They're skipped without any message as to why.
I'm not familiar at all with py.test, maybe someone who knows what he's doing will have more luck.
I'm pretty sure it's some silly mistake on my part.

**Comment by graingert** [comment]
ah that's cool, sqlaclhemy fires off a query that's nice.

I hope that works atomically :P

**Comment by mtambos** [comment]
> I hope that works atomically :PI guess that depends on the transaction's isolation level.

**Comment by graingert** [comment]
@mtambos can you resolve all the issues in the "Todos" section

**Comment by mtambos** [comment]
Sorry, I forgot to answer this.
I could fix everything except the tests in the next couple weeks.
As I mentioned in a previous post, I wrote a couple tests, but I could't get them to work.

File: tests/test_postfetch_lastrowid.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class Foo(Base):
+    __tablename__ = 'foo'
+
+    id = Column(Integer, primary_key=True, info={'identity': (0, 1)})
+    some_int = Column(Integer)
+
+
+def test_postfetch_lastrowid_false(_redshift_database_tool):
```
Comment (graingert) [inline tests/test_postfetch_lastrowid.py]: Don't include this test.

File: tests/test_postfetch_lastrowid.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        for i in range(5):
+            session.add(Foo(some_int=i))
+        foo2 = session.query(Foo).filter(Foo.some_int == 4).first()
+        assert foo2.id > 0
+    finally:
+        Base.metadata.drop_all(redshift_engine)
+
+
+def test_postfetch_lastrowid_true(_redshift_database_tool):
+    RedshiftDialect.postfetch_lastrowid = True
```
Comment (graingert) [inline tests/test_postfetch_lastrowid.py]: postfetch_lastrowid is already True.

File: tests/test_postfetch_lastrowid.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        for i in range(5):
+            session.add(Foo(some_int=i))
+        foo2 = session.query(Foo).filter(Foo.some_int == 4).first()
+        assert foo2.id > 0
+    finally:
+        Base.metadata.drop_all(redshift_engine)
+
+
+def test_postfetch_lastrowid_true(_redshift_database_tool):
```
Comment (graingert) [inline tests/test_postfetch_lastrowid.py]: Create your Foo model in  rs_sqla_test_utils/models as PostfetchExample.
Then:
```python
from rs_sqla_test_utils import models

def test_postfetch_lastrowid(redshift_session):
    session = redshift_session
    for i in range(5):
        postfetchExample = models.PostfetchExample(some_int=i)
        session.add(postfetchExample)
    assert postfetchExample.id == 4
```

File: tests/test_postfetch_lastrowid.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+Base = declarative_base()
+
+
+class Foo(Base):
+    __tablename__ = 'foo'
+
+    id = Column(Integer, primary_key=True, info={'identity': (0, 1)})
+    some_int = Column(Integer)
```
Comment (graingert) [inline tests/test_postfetch_lastrowid.py]: you don't really need this some_int column.

**Comment by graingert** [comment]
@mtambos please add the changelog, assert the license etc.

**Comment by jseabold** [comment]
What's still needed here a line added to the changelog?

**Comment by graingert** [comment]
@jseabold 
todos

**Comment by jseabold** [comment]
Yep I saw that. I guess I'm asking for clarification (on the author's behalf) and didn't know when that was created vs. the activity that followed.

What does MIT compatible mean? The author needs to positively assert the license?

There seem to be tests. Do the remaining comments need to be addressed?

@mtambos Could you add a line to the docs and update CHANGES.rst. Would be great to get this merged. I'm maintaining a fork with this in it right now.

**Comment by graingert** [comment]
@jseabold you're welcome to make another PR.

**Comment by graingert** [comment]
@jseabold currently everything is outstanding, and most importantly the author has not yet allowed me to include this code in an MIT licensed product.

**Comment by graingert** [comment]
closing in favour of #91

**Comment by mtambos** [comment]
Sorry I was out on vacation and completely forgot about this.
Do you still need me to do anything?

**Comment by graingert** [comment]
If you can find out why this change broke master on Travis, I can get a
release out. It looks like the IDs coming back are all 0s

**Pull request closed without merging**